### ISSUE TITLE
Add more examples to more str functions

### ISSUE THREAD

**Issue by SoniEx2** [rust-lang/rust#54520]
Functions like contains, starts_with, ends_with, match_indices, etc could use more examples, for example `foo.contains('\n')` (check if a string contains newlines).

find, split, matches, trim_matches, etc all have them.

**Comment by hellow554** [comment]
Why don't you submit a PR ;) It's a very good and easy topic to work on

**Comment by SoniEx2** [comment]
it shows you can use single characters and stuff?

**Comment by shepmaster** [comment]
Do you also suggest adding examples using a closure? 

I think it would be better to highlight that these functions take a [`Pattern`](https://doc.rust-lang.org/std/str/pattern/trait.Pattern.html), link to that documentation, and show examples of all the [standard library provided implementations](https://doc.rust-lang.org/std/str/pattern/trait.Pattern.html#implementors). This reduces redundancy.

**Comment by shepmaster** [comment]
Also worth noting that `Pattern` is unstable and there's an [RFC out there to improve it](https://github.com/rust-lang/rfcs/pull/2500) /cc @kennytm .

**Comment by frewsxcv** [comment]
> it shows you can use single characters and stuff?

the documentation for this is in [`Pattern`](https://doc.rust-lang.org/std/str/pattern/trait.Pattern.html). extending the docs for all APIs that use `Pattern` demonstrating all the possible `Pattern` variations seems unrelated to the APIs we're documenting, in my opinion.

**Comment by pm215** [comment]
The most salient feature of the Pattern documentation is all the "nightly only", "nightly only" warnings in it. I think it would be better to be able to understand the documentation for non-nightly features without having to look at documentation that is so marked. Also in general, if you're relatively new to Rust, I think it's much easier to understand a concrete example than to follow a link to a page documenting an abstract Trait and then have to figure out how the trait applies to the particular situation you were interested in. More examples in the Pattern page itself would probably help too.

**Comment by hellow554** [comment]
https://doc.rust-lang.org/std/primitive.str.html#method.split

> The pattern can be a &str, char, or a closure that determines the split.
### Examples
  * Simple patterns:
  * A more complex pattern, using a closure:
  * If a string contains multiple contiguous separators, you will end up with empty strings in the output:
  * Contiguous separators are separated by the empty string.
  * ...

I honstely think there are more than enough code examples all with explanation. If you think we are missing something, please feel free to suggest one or open a PR and add one. We are welcoming it!

**Comment by pm215** [comment]
Yes, the split method has a good set of examples. I'm here because I started at the trim_start_matches method, which doesn't so much. I raised a bug (﻿﻿﻿56891); it was merged with this one; estebank filed a PR 56898 which I think would be fine as a fix for it; shepmaster commented on that PR saying they were against merging it; so I'm here trying to argue the case for being a bit more verbose with the examples.

**Comment by hellow554** [comment]
@pm215 pro tip: You can put a `#` in front of the number and the PR/issue will be linked directly ;)

**Comment by shepmaster** [comment]
> shepmaster commented on that PR saying they were against merging it

To be clear, I'm only one voice; others may (and probably do) disagree with me. 

I'm mostly arguing from a maintenance perspective; this request is the equivalent of "inline the `Pattern` docs into every API function that takes a `Pattern` argument". We don't do that for `Iterator`, a much more common argument (I believe). It makes it harder to explain things once and keep something up to date. Once `Pattern` stabilizes, any crate can implement it (think `Regex`), but cannot add itself to the `str::split` documentation after the fact.

**Comment by SoniEx2** [comment]
The problem is that Pattern is nightly-only, and the docs aren't written to reflect Pattern matching but instead give examples based on concrete impls.

**Comment by SoniEx2** [comment]
All of those cases are used once or only a few times, and as such are allowed to rely on concrete stuff for the docs.

The problem with Pattern is that it's used MANY times...

Hmm. I wonder if we could boil everything down to one or two core Pattern-using things and describe everything else as a combination of them, so that inlining Pattern's docs only needs to happen a handful of times...

**Comment by DevQps** [comment]
> > shepmaster commented on that PR saying they were against merging it
> 
> To be clear, I'm only one voice; others may (and probably do) disagree with me.
> 
> I'm mostly arguing from a maintenance perspective; this request is the equivalent of "inline the `Pattern` docs into every API function that takes a `Pattern` argument". We don't do that for `Iterator`, a much more common argument (I believe). It makes it harder to explain things once and keep something up to date. Once `Pattern` stabilizes, any crate can implement it (think `Regex`), but cannot add itself to the `str::split` documentation after the fact.

Personally I think that making users aware that a `Pattern` is passed is the right way. `Pattern` is pretty powerful, and once people realize that everything that implements `Pattern` (such as `&str` and `char`) everything will probably become much more clear for them without making it seem like it only works on `&str` and `char`s.

Maybe we can find some middle ground by linking the examples to Pattern more clearly but also stating that Pattern is implemented by `char`s and `&str`? That would probably only add a single line of clarification without having to provide many examples all over the place. I am curious to what you think of this! I don't mind picking this up, if @SoniEx2 don't want to! I probably need some extra pointers on which docs to edit however :)

**Comment by shepmaster** [comment]
Also be aware of [RFC 2005](https://rust-lang.github.io/rfcs/2500-needle.html), which enhances `Pattern`and renames it to `Needle`.

**Comment by DevQps** [comment]
@shepmaster Thanks for the advice, do you think it would be wise to wait with writing any documentation as that might soon be replaced with the Needle API? Or do you suspect that that will still take a while?

**Comment by DevQps** [comment]
@shepmaster I hope you're still willing to respond to my previous comment here! :) Then I can either start writing some documentation or we can maybe wait for Needle to come through. (Y)

**Comment by shepmaster** [comment]
> wait with writing any documentation as that might soon be replaced with the Needle API

I don't expect it to be anytime soon — @kennytm do you have any feelings about when it might occur?
